chain/indexer/src/streamer/mod.rs

```diff
@@ -149,34 +163,65 @@ pub async fn build_streamer_message(
-                        prev_block,
+                // Attempt to find the receipt in the cache
+                if let Some(receipt) = DELAYED_LOCAL_RECEIPTS_CACHE
+                    .write()
+                    .unwrap() // TODO handle this unwrap
+                    .remove(&execution_outcome.id)
+                {
+                    receipt
+                } else {
+                    tracing::warn!(
+                        target: INDEXER,
+                        "Receipt {} is missing in block and in DELAYED_LOCAL_RECEIPTS_CACHE, starting to look for it in up to 1000 blocks back in time",
                         execution_outcome.id,
-                    )
-                    .await?
-                    {
-                        break 'find_local_receipt receipt;
-                    }
+                    );
+                    // Receipt might be missing only in case of delayed local receipt
+                    // that appeared in some of the previous blocks
+                    // we will be iterating over previous blocks until we found the receipt
+                    let mut prev_block_tried = 0u16;
+                    let mut prev_block_hash = block.header.prev_hash;
+                    'find_local_receipt: loop {
+                        if prev_block_tried > 1000 {
+                            panic!("Failed to find local receipt in 1000 prev blocks");
+                        }
+                        // Log a warning every 100 blocks
+                        if prev_block_tried % 100 == 0 {
+                            tracing::warn!(
+                                target: INDEXER,
+                                "Still looking for receipt {} in previous blocks. {} blocks back already",
+                                execution_outcome.id,
+                                prev_block_tried,
+                            );
+                        }
+                        let prev_block = match fetch_block(&client, prev_block_hash).await {
+                            Ok(block) => block,
+                            Err(err) => panic!("Unable to get previous block: {:?}", err),
+                        };
+
+                        prev_block_hash = prev_block.header.prev_hash;
+
+                        if let Some(receipt) = find_local_receipt_by_id_in_block(
+                            &client,
+                            &runtime_config,
+                            prev_block,
+                            execution_outcome.id,
+                        )
+                        .await?
+                        {
+                            tracing::debug!(
+                                target: INDEXER,
+                                "Found receipt {} in previous block {}",
+                                execution_outcome.id,
+                                prev_block_tried,
+                            );
+                            metrics::LOCAL_RECEIPT_LOOKUP_IN_HISTORY_BLOCKS_BACK
```

Comment:
This metric in Grafana will allow us to see the chart of how far we went in history if we had to find the receipt.